Clarify that extra properties are not allowed.

garemoko opened the pull request:
Fixes https://github.com/adlnet/xAPI-Spec/issues/335

I'm sure there'll be some debate on my use of MUST, SHOULD* and MAY* here.

garemoko commented on xAPI.md:
```diff
@@ -1884,6 +1884,8 @@ Complete IRI validation is extremely difficult, so much of the burden for ensuri
 * Values SHOULD be considered to be case sensitive unless specified otherwise. 
 * Lowercase SHOULD be used to send case insensitive data.
+* Additional properties SHOULD* NOT be added to statements and other objects unless explicitly allowed and the 
+LRS MAY* reject statements and objects containing such additional properties.
```
Note the use of "other objects" here, so the specific requirements around About and Person are just there for clarity.

fugu13 replied:
It might be good to add a note to APs that there are definitely LRSs that do so. The other objects bit is a bit weird since it isn't possible to set any other objects (that have a spec-restricted property list) with the spec -- about and person objects are only returned by the LRS.

fugu13 replied:
(the point of the AP note is that if you want to send interoperable statements, you definitely can't include any additional properties)

garemoko replied:
Other objects relates to the LRS. I should make that clearer.

garemoko commented on xAPI.md:
````diff
@@ -3352,9 +3356,12 @@ Returns: ```200 OK```, Person Object
 same definition as the similarly named property from Agent Objects.  
 
+* Additional properties not listed here SHOULD* NOT be added to this object and each 
+property MUST occur only once.  
````
MUST for multiplicity because this could cause chaos, nobody is doing it, most JSON implementations don't support it and whilst the JSON spec doesn't specifically forbid it, it doesn't explicitly allow it either. 

SHOULD* for additional properties for the same reason as statements above. I'd be ok with a MUST too though.

garemoko commented on xAPI.md:
```diff
@@ -3452,6 +3459,8 @@ the latest minor and patch version the LRS conforms to, for each major version.
     "0.9" and "0.95" MAY be included. (For the purposes of this requirement, "0.9" and "0.95"
     are considered major versions.)
+* Additional properties MUST NOT be added to this object outside of extensions and each 
+property MUST occur only once.  
```
MUST for multiplicity as above. Here I'm confident with a MUST for additional properties too as we explicitly say on [line 3457](https://github.com/adlnet/xAPI-Spec/pull/580/files#diff-68cd6732e400b6e64376effdb84b970eR3457) ```An LRS MUST return the JSON document described above```.

garemoko commented:
Updated following yesterday's call. I'm frustrated that there's nowhere central for these requirements so have raised https://github.com/adlnet/xAPI-Spec/issues/590. I'd like to see this merged first though.

andyjohnson commented:
I almost want to have a section at the very beginning describing the basics of a JSON-binding of xAPI.  I think you can use all MUSTS where warranted because the existence of the tables and extensions indicates the intentions of the spec.

andyjohnson commented:
How do others feel about changing the SHOULD * NOT to MUST?  It already has an "unless" in the requirement.

garemoko commented:
There's two SHOULD* NOTs - which one do you mean, or do you mean both? I'm not against making either/both of them MUST.

fugu13 commented on xAPI.md:
```diff
@@ -1893,9 +1895,11 @@ Complete IRI validation is extremely difficult, so much of the burden for ensuri
-      an enumerated value given in the standard exactly.
+      an enumerated value given in this specification exactly.
+    * where a key or value is not allowed by this specification.
+    * where a key occurs mutliple times within an object. 
```
minor typo, "mutliple"

garemoko commented:
Changes based on @fugu13's feedback. The requirements about additional properties in other objects and the LRS rejecting statements with additional properties are now in the LRS Requirements section and I added a note in the client requirements section flagging up that LRS are recommended to reject statements with additional properties. 

I also changed the MAY* to SHOULD* requiring the LRS to reject statements with additional properties. SHOULD* is no worse that MAY* for clients already using additional properties, but it's better in preventing future clients from using additional properties. 

Effectively this is strongly and clearly recommending that no additional properties are added to statements, but quietly allowing the possibly (until 1.1) that an LRS and client could agree to include/accept additional statement properties. I think that correctly represents the current ambiguous situation and sets us up well to close it in 1.1.

garemoko commented:
Now also fixes https://github.com/adlnet/xAPI-Spec/issues/597 which is an unrelated, but very minor markdown change that didn't need its own PR.

andyjohnson commented:
I'd like to recommend tweaking the note:  

"Note that the LRS is recommended to reject Statements containing additional properties, and many do. Additional properties in Statements will mean that the statement will not be interoperable with all LRS, hence the strong recommendation above not to include additional properties in Statements. "

down to:

"Note: The LRS is recommended to reject Statements containing additional properties. Additional properties in Statements would mean that the Statement would not be interoperable with all LRS"

1) I don't think we should use what is in the wild as evidence for anything in the spec.  We can make decisions TO the spec based on them, but to reference them isn't something we've done before.  I also don't think the "hence" clause adds any value, and we should be trying to make this as succinct as we have been.  Remember too we are changing this requirement to a MUST in the future, so let's limit the test and references and why I changed the "will" to "would", I think we should assume they are going to implement our recommendations!

garemoko commented:
I'm happy with that change, but the wording was added in response to [this comment](https://github.com/adlnet/xAPI-Spec/pull/580#discussion_r22745661) from @fugu13 so it'd be good to hear his response to your suggestion before I make the change.

fugu13 commented:
I'm fine with @andyjohnson's proposed change, I think it has the needed effect more succinctly.

garemoko commented:
Done :-)

andyjohnson commented:
I think merging the other PR this morning knocked this one out of auto-merge.  @garemoko can you take a quick peek?

garemoko commented:
On it.

garemoko commented:
That should be good now. Apologies for the two extra commits.
